File: backend/app/api/apps.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException

from backend.app.schemas import AppAutostartUpdateRequest, AppConfigUpdateRequest, AppInstallRequest, AppUninstallRequest
from backend.app.services import apps_service
from backend.app.services.auth import auth_token

router = APIRouter(prefix="/api/apps", tags=["apps"])


@router.post("/repo/sync")
async def sync_repo(token: str) -> dict[str, Any]:
    await auth_token(token)
    try:
        return apps_service.sync_catalog_from_repo()
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"同步远端仓库失败: {exc}") from exc


@router.get("/catalog")
async def get_catalog(token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.get_catalog()


@router.get("/installed")
async def get_installed(token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.get_installed()


@router.post("/{app_id}/install")
async def install_app(app_id: str, payload: AppInstallRequest, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.create_install_task(app_id, payload.install_runtime)


@router.get("/tasks/{task_id}")
async def get_task(task_id: str, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.app_task_detail(task_id)


@router.get("/{app_id}/status")
async def get_status(app_id: str, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.app_runtime_detail(app_id)


@router.post("/{app_id}/start")
async def start_app(app_id: str, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.start_app_runtime(app_id)


@router.post("/{app_id}/stop")
async def stop_app(app_id: str, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.stop_app_runtime(app_id)


@router.post("/{app_id}/autostart")
async def set_autostart(app_id: str, payload: AppAutostartUpdateRequest, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.set_autostart(app_id, payload.enabled)


@router.post("/{app_id}/uninstall")
async def uninstall_app(app_id: str, payload: AppUninstallRequest, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.uninstall_app(app_id, payload.purge)


@router.get("/{app_id}/config")
async def get_app_config(app_id: str, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.get_app_config(app_id)


@router.put("/{app_id}/config")
async def put_app_config(app_id: str, payload: AppConfigUpdateRequest, token: str) -> dict[str, Any]:
    await auth_token(token)
    return apps_service.put_app_config(app_id, payload.version, payload.data)

File: frontend/src/services/robot/apps.js
import { apiGet, apiPost, apiPut } from './api'
import { robotState } from './state'

export async function refreshApps() {
  if (!robotState.token) return
  robotState.appsError = ''
  try {
    await apiPost('/api/apps/repo/sync')
    const [catalogResp, installedResp] = await Promise.all([
      apiGet('/api/apps/catalog'),
      apiGet('/api/apps/installed'),
    ])
    robotState.appCatalog = catalogResp.items || []
    robotState.installedApps = installedResp.items || []
  } catch (err) {
    robotState.appsError = err.message
  }
}

export async function installApp(appId, installRuntime) {
  const result = await apiPost(`/api/apps/${encodeURIComponent(appId)}/install`, {
    install_runtime: Boolean(installRuntime),
  })
  const taskId = result.task_id
  if (taskId) {
    await pollTaskUntilDone(taskId)
  }
  await refreshApps()
}

export async function startApp(appId) {
  await apiPost(`/api/apps/${encodeURIComponent(appId)}/start`)
  await refreshApps()
}

export async function stopApp(appId) {
  await apiPost(`/api/apps/${encodeURIComponent(appId)}/stop`)
  await refreshApps()
}

export async function setAppAutostart(appId, enabled) {
  await apiPost(`/api/apps/${encodeURIComponent(appId)}/autostart`, { enabled: Boolean(enabled) })
  await refreshApps()
}

export async function uninstallApp(appId, purge = false) {
  await apiPost(`/api/apps/${encodeURIComponent(appId)}/uninstall`, { purge: Boolean(purge) })
  await refreshApps()
}

export async function getAppConfig(appId, forceReload = false) {
  if (!forceReload && robotState.appConfigById[appId]) {
    return robotState.appConfigById[appId]
  }
  const cfg = await apiGet(`/api/apps/${encodeURIComponent(appId)}/config`)
  robotState.appConfigById = {
    ...robotState.appConfigById,
    [appId]: cfg,
  }
  return cfg
}

export async function saveAppConfig(appId, version, data) {
  const resp = await apiPut(`/api/apps/${encodeURIComponent(appId)}/config`, { version, data })
  const cfg = await getAppConfig(appId, true)
  await refreshApps()
  return { version: resp.version, config: cfg }
}

export async function pollTaskUntilDone(taskId) {
  const maxTicks = 120
  for (let i = 0; i < maxTicks; i += 1) {
    const task = await apiGet(`/api/apps/tasks/${encodeURIComponent(taskId)}`)
    robotState.appTaskById = { ...robotState.appTaskById, [taskId]: task }
    if (task.status === 'succeeded') return task
    if (task.status === 'failed') {
      throw new Error(task.error || task.message || '安装失败')
    }
    await new Promise((resolve) => setTimeout(resolve, 800))
  }
  throw new Error('安装超时，请稍后刷新查看')
}
